vnext/Microsoft.ReactNative/Views/TouchEventHandler.cpp

```diff
@@ -449,49 +469,76 @@ bool TouchEventHandler::TagFromOriginalSource(
+  // The character is before the point if the Y-coordinate of the point is
+  // below (greater than) the bottom of the character rect, or if the
+  // Y-coordinate is below (greater than) the top of the character rect and
+  // the X-coordinate is greater than the right side of the character rect:
+  // ┌───────────┐
+  // │     ┌─────┘
+  // └─────┘ (x,y)
+  return point.Y > bottom || (point.Y > rect.Y && point.X > right);
+}
 
-      if (isHit) {
-        tag = el.GetValue(xaml::FrameworkElement::TagProperty()).try_as<winrt::IPropertyValue>();
-        if (tag) {
-          return tag;
-        }
-      }
-    } else if (const auto run = el.try_as<xaml::Documents::Run>()) {
-      const auto start = el.ContentStart();
-      const auto end = el.ContentEnd();
-
-      auto startRect = start.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
-      auto endRect = end.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
-
-      // Swap rectangles in RTL scenarios.
-      if (startRect.X > endRect.X) {
-        const auto tempRect = startRect;
-        startRect = endRect;
-        endRect = tempRect;
-      }
+bool IsCharacterAfter(winrt::TextPointer textPointer, winrt::Point point) {
+  const auto rect = textPointer.GetCharacterRect(winrt::LogicalDirection::Forward);
+  const auto bottom = rect.Y + rect.Height;
+
+  // The character is after the point if the Y-coordinate of the point is above
+  // (less than) the top of the character rect, or if the Y-coordinate is above
+  // (less than) the bottom of the character rect and the X-coordinate is less
+  // than the left side of the character rect:
+  //
+  // (x,y) ┌─────┐
+  // ┌─────┘     │
+  // └───────────┘
+  return point.Y < rect.Y || (point.Y < bottom && point.X < rect.X);
+}
 
-      // Approximate the bounding rect (for now, don't account for text wrapping).
-      if ((startRect.X <= pointerPos.X) && (endRect.X + endRect.Width >= pointerPos.X) &&
-          (startRect.Y <= pointerPos.Y) && (endRect.Y + endRect.Height >= pointerPos.Y)) {
-        isHit = true;
-        return nullptr;
-      }
+winrt::TextPointer TouchEventHandler::GetPositionFromPoint(winrt::TextBlock textBlock, winrt::Point point) {
+  // Since characters in a TextBlock are sorted from top-left to bottom-right,
+  // we can use binary search to find the character whose bounds contains the
+  // pointer point.
+  //
+  // This algorithm currently makes the following assumptions:
+  // 1. Characters on the same line have the same Rect::X value
+  // 2. The text is left-to-right
+  auto textPointer = textBlock.ContentStart();
+  auto L = 0;
+  auto R = /* n - 1 */ textBlock.ContentEnd().Offset();
+  while (L <= R) {
+    const auto m = /* floor */ (L + R) / 2;
+    const auto relativeOffset = m - textPointer.Offset();
+    textPointer = textPointer.GetPositionAtOffset(relativeOffset, winrt::LogicalDirection::Forward);
+    if (IsCharacterBefore(textPointer, point) /* A[m] < T */) {
```

Comment:
I think this name should be something related to 'after/following' - in bin search you move the lower band if target is past the median. Likewise for @534

Reply:
Fixed.